core-bundle/src/EventListener/DataContainer/Undo/UndoListenerTrait.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DataContainer\Undo;
+
+use Contao\Controller;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Doctrine\DBAL\Connection;
+
+/**
+ * @internal
+ */
+trait UndoListenerTrait
+{
+    private Connection $connection;
+    private ContaoFramework $framework;
+
+    private function getParentTableForRow(string $table, array $data): ?array
+    {
+        if (isset($GLOBALS['TL_DCA'][$table]['config']['dynamicPtable']) && true === $GLOBALS['TL_DCA'][$table]['config']['dynamicPtable']) {
+            return ['table' => $data['ptable'], 'id' => $data['pid']];
+        }
+
+        if (isset($GLOBALS['TL_DCA'][$table]['config']['ptable'])) {
+            return ['table' => $GLOBALS['TL_DCA'][$table]['config']['ptable'], 'id' => $data['pid']];
+        }
+
+        return null;
+    }
+
+    private function getTranslatedTypeFromTable(string $table): string
+    {
+        /** @var Controller $controller */
+        $controller = $this->framework->getAdapter(Controller::class);
+        $controller->loadLanguageFile($table);
+
+        return isset($GLOBALS['TL_LANG'][$table]['_type']) ? $GLOBALS['TL_LANG'][$table]['_type'][0] : $table;
```

Comment:
`['_type'][0]` is defined in the language files but seems unused? I have two thoughts about that

1. this basically is a label for each table, right? I wonder if we don't have that already, if we really need it and if we could use that in other places. How do we for example generate the breadcrumb?
2. By having these labels in the default language file (e.g. `$GLOBALS['TL_LANG']['MODEL'][...]`) we could prevent the system from needing to load all table translation files just for this.

Reply:
Sadly, there are no translations for the table names yet.

I agree that these would be useful in multiple places! We often customize the version overview on the backend dashboard to show translated table names instead of `tl_content`, `tl_module` etc. to the editor.

The translations should go into `modules.xlf` or alternatively into `default.xlf`, possibly as something like `tl_lang.table.[table]` ( e.g. `tl_lang.table.tl_content = 'Content element'`).

Since I assume there'll be some discussion about the correct names for some tables, maybe these should get their own pull request to keep the focus. What do you think @bezin?

If we have these labels, I'll happily provide another pull request to use them in the version overview.

Reply:
Hello all, we discussed this in a Mumble call two and a half years ago :D We decided we want to introduce singular and plural translations for tables/models/types which is what I implemented here. 

I agree that it might be an issue when we put it into each individual DCA translation file. `modules.xlf` is IMHO the wrong place (a module, back-end or front-end can deal with many models),, but `default.xlf` sounds suitable.

@aschempp You are right, they do exist for some tables, but not in a universal way. Have a look at the form or comments backend module in the demo: If you edit a form, the German breadcrumb is `Formulare > Felder`, but I would need `Formular` and `Feld`. In the comments module, you only have `Kommentare`, but no singular either. 

That said, I still think it is a good idea and we should introduce these in a universal way.

I think, for the core elements the translation is a no brainer, as we use singular and plural labels in many places; we simply lack these labels as separately translated strings. As one of my goals with this PR was **to make the undo module less technically**, I would not like to merge this without translated table names (for the core types), but if this causes discussion I am happy to provide a separate PR to merge the translations first.

Reply:
Thanks for the background information! 

I hadn't thought of the plural forms - they would also come in handy in many places. Nice!

If the translations are collected in a single file, they could probably also be discussed here if necessary.

Reply:
I moved all translations to `contao_default.TABLES` Documentation needs an udpate here then: https://docs.contao.org/dev/framework/translations/#categories